File: impl/ocaml/src/theme_functions.ml
(* Theme function substitution (SPEC §6.11). *)

open Ast
open Utils

let function_names = [ "--spacing("; "--alpha("; "--theme("; "theme(" ]

(* Whether a value contains a theme function call. *)
let has_theme_function value =
  List.exists
    (fun name ->
      let rec find from =
        match index_of (after value from) name with
        | -1 -> false
        | i ->
            let i = from + i in
            if i = 0 || not (is_ident_byte value.[i - 1]) then true else find (i + 1)
      in
      find 0)
    function_names

let function_args nodes = List.map String.trim (segment (Value.to_css nodes) ',')

let spacing_function nodes (ds : Design_system.t) =
  let args = function_args nodes in
  (match args with
  | [ a ] when a <> "" -> ()
  | _ ->
      Twill_error.failf "The --spacing(…) function requires exactly one argument, but received %d."
        (List.length (List.filter (fun a -> a <> "") args)));
  let multiplier =
    match Theme.resolve ds.theme None [ "--spacing" ] 0 with
    | Some m -> m
    | None -> Twill_error.fail "The --spacing(…) function requires that the `--spacing` theme variable exists, but it was not found."
  in
  match List.hd args with "0" -> "0px" | "1" -> multiplier | arg -> "calc(" ^ multiplier ^ " * " ^ arg ^ ")"

let alpha_function nodes =
  let text = Value.to_css nodes in
  if List.length (segment text ',') <> 1 then
    Twill_error.failf "The --alpha(…) function requires exactly one argument in the form `<color> / <alpha>`, but received `%s`." text;
  match segment text '/' with
  | [ color; alpha ] when String.trim color <> "" && String.trim alpha <> "" -> Theme.with_alpha (String.trim color) (String.trim alpha)
  | _ -> Twill_error.failf "The --alpha(…) function requires a color and an alpha value separated by `/`, but received `%s`." text

let is_theme_reference name = name = "var" || name = "theme" || name = "--theme"

let rec inject_fallback_nodes (nodes : Value.vnode list) fallback =
  List.exists
    (fun n ->
      match n with
      | Value.Fn f ->
          if inject_fallback_nodes f.fnodes fallback then true
          else if not (is_theme_reference f.fname) then false
          else begin
            let rec find i = function
              | [] -> -1
              | Value.Sep "," :: _ -> i
              | _ :: rest -> find (i + 1) rest
            in
            match find 0 f.fnodes with
            | -1 ->
                f.fnodes <- f.fnodes @ [ Value.sep ","; Value.sep " "; Value.word fallback ];
                true
            | comma ->
                let before = List.filteri (fun i _ -> i <= comma) f.fnodes in
                let after_comma = List.filteri (fun i _ -> i > comma) f.fnodes in
                if String.trim (Value.to_css after_comma) = "initial" then begin
                  f.fnodes <- before @ [ Value.sep " "; Value.word fallback ];
                  true
                end
                else false
          end
      | _ -> false)
    nodes

(* Injects [fallback] into the innermost var(...), theme(...), or
   --theme(...) call that has no fallback or whose fallback is `initial`. *)
let inject_fallback value fallback =
  let ast = Value.parse value in
  ignore (inject_fallback_nodes ast fallback);
  Value.to_css ast

let theme_function nodes (ds : Design_system.t) in_at_rule =
  let args = function_args nodes in
  let key = match args with k :: _ -> k | [] -> "" in
  let inline, key =
    if has_suffix key " inline" then (true, String.trim (trim_suffix key " inline")) else (in_at_rule, key)
  in
  let has_fallback = List.length args > 1 in
  let fallback = if has_fallback then String.concat ", " (List.tl args) else "" in
  if not (has_prefix key "--") then
    Twill_error.failf "The --theme(…) function can only be used with CSS variables from your theme, but received `%s`." key;
  match Theme.resolve_theme_value ds.theme key inline with
  | None ->
      if has_fallback then fallback else Twill_error.failf "Could not resolve value for theme function: `--theme(%s)`." key
  | Some resolved ->
      if (not has_fallback) || fallback = "initial" then resolved
      else if resolved = "initial" then fallback
      else if has_prefix resolved "var(" || has_prefix resolved "theme(" || has_prefix resolved "--theme(" then
        inject_fallback resolved fallback
      else resolved

let legacy_theme_function nodes (ds : Design_system.t) =
  let args = function_args nodes in
  let key = match args with k :: _ -> k | [] -> "" in
  let key = match unquote key with Some k -> k | None -> key in
  match Theme.resolve_theme_value ds.theme key true with
  | Some resolved -> resolved
  | None ->
      if List.length args > 1 then String.concat ", " (List.tl args)
      else Twill_error.failf "Could not resolve value for theme function: `theme(%s)`." key

(* Substitutes theme functions in one value. *)
let substitute_in_value value ds in_at_rule =
  let ast = ref (Value.parse value) in
  Value.walk ast (fun n _ ->
      match n with
      | Value.Fn f -> (
          let result =
            match f.fname with
            | "--spacing" -> Some (spacing_function f.fnodes ds)
            | "--alpha" -> Some (alpha_function f.fnodes)
            | "--theme" -> Some (theme_function f.fnodes ds in_at_rule)
            | "theme" -> Some (legacy_theme_function f.fnodes ds)
            | _ -> None
          in
          match result with Some r -> Value.VReplace [ Value.word r ] | None -> Value.VContinue)
      | _ -> Value.VContinue);
  Value.to_css !ast

let at_rules_with_params = [ "@media"; "@custom-media"; "@container"; "@supports" ]

(* Substitutes theme functions in declaration values and in the params of
   @media, @custom-media, @container, and @supports. Returns the features. *)
let substitute_functions (ast : nodes) ds =
  let features = ref 0 in
  ignore
    (walk ast (fun n _ ->
         (match n with
         | Declaration d when (not d.no_value) && has_theme_function d.value ->
             d.value <- substitute_in_value d.value ds false;
             features := !features lor Features.theme_function
         | At_rule at when List.mem at.name at_rules_with_params && has_theme_function at.params ->
             at.params <- substitute_in_value at.params ds true;
             features := !features lor Features.theme_function
         | _ -> ());
         Continue));
  !features

File: impl/go/theme_functions.go
package twill

import (
	"regexp"
	"strings"
)

var themeFunctionPattern = regexp.MustCompile(`(^|[^\w-])(?:--spacing|--alpha|--theme|theme)\(`)

// HasThemeFunction reports whether a value contains a theme function call.
func HasThemeFunction(value string) bool { return themeFunctionPattern.MatchString(value) }

var atRulesWithParams = map[string]bool{"@media": true, "@custom-media": true, "@container": true, "@supports": true}

// SubstituteFunctions substitutes theme functions in declaration values and
// in the params of @media, @custom-media, @container, and @supports
// (SPEC §6.11).
func SubstituteFunctions(ast *[]Node, ds *DesignSystem) (Features, error) {
	var features Features
	var failure error
	Walk(ast, func(n Node, _ *WalkUtils) WalkAction {
		switch n := n.(type) {
		case *Declaration:
			if !n.NoValue && HasThemeFunction(n.Value) {
				value, err := SubstituteFunctionsInValue(n.Value, ds, false)
				if err != nil {
					failure = err
					return Stop
				}
				n.Value = value
				features |= FeatureThemeFunction
			}
		case *AtRule:
			if atRulesWithParams[n.Name] && HasThemeFunction(n.Params) {
				params, err := SubstituteFunctionsInValue(n.Params, ds, true)
				if err != nil {
					failure = err
					return Stop
				}
				n.Params = params
				features |= FeatureThemeFunction
			}
		}
		return Continue
	})
	return features, failure
}

// SubstituteFunctionsInValue substitutes theme functions in one value.
func SubstituteFunctionsInValue(value string, ds *DesignSystem, inAtRule bool) (string, error) {
	ast := ParseValue(value)
	var failure error
	WalkValue(&ast, func(n ValueNode, _ *ValueFunction) ([]ValueNode, bool, bool) {
		f, ok := n.(*ValueFunction)
		if !ok || failure != nil {
			return nil, false, false
		}
		var result string
		var err error
		switch f.Name {
		case "--spacing":
			result, err = spacingFunction(f.Nodes, ds)
		case "--alpha":
			result, err = alphaFunction(f.Nodes)
		case "--theme":
			result, err = themeFunction(f.Nodes, ds, inAtRule)
		case "theme":
			result, err = legacyThemeFunction(f.Nodes, ds)
		default:
			return nil, false, false
		}
		if err != nil {
			failure = err
			return nil, false, true
		}
		return []ValueNode{Word(result)}, true, true
	})
	if failure != nil {
		return "", failure
	}
	return ValueToCSS(ast), nil
}

func functionArgs(nodes []ValueNode) []string {
	parts := Segment(ValueToCSS(nodes), ',')
	for i, p := range parts {
		parts[i] = strings.TrimSpace(p)
	}
	return parts
}

func spacingFunction(nodes []ValueNode, ds *DesignSystem) (string, error) {
	args := functionArgs(nodes)
	if len(args) != 1 || args[0] == "" {
		count := 0
		for _, a := range args {
			if a != "" {
				count++
			}
		}
		return "", Errorf("The --spacing(…) function requires exactly one argument, but received %d.", count)
	}
	multiplier, ok := ds.Theme.Resolve(nil, []string{"--spacing"}, 0)
	if !ok {
		return "", Errorf("The --spacing(…) function requires that the `--spacing` theme variable exists, but it was not found.")
	}
	switch args[0] {
	case "0":
		return "0px", nil
	case "1":
		return multiplier, nil
	}
	return "calc(" + multiplier + " * " + args[0] + ")", nil
}

func alphaFunction(nodes []ValueNode) (string, error) {
	text := ValueToCSS(nodes)
	if len(Segment(text, ',')) != 1 {
		return "", Errorf("The --alpha(…) function requires exactly one argument in the form `<color> / <alpha>`, but received `%s`.", text)
	}
	parts := Segment(text, '/')
	if len(parts) != 2 || strings.TrimSpace(parts[0]) == "" || strings.TrimSpace(parts[1]) == "" {
		return "", Errorf("The --alpha(…) function requires a color and an alpha value separated by `/`, but received `%s`.", text)
	}
	return WithAlpha(strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])), nil
}

func themeFunction(nodes []ValueNode, ds *DesignSystem, inAtRule bool) (string, error) {
	args := functionArgs(nodes)
	key := ""
	if len(args) > 0 {
		key = args[0]
	}
	inline := inAtRule
	if strings.HasSuffix(key, " inline") {
		inline = true
		key = strings.TrimSpace(key[:len(key)-len(" inline")])
	}
	fallback := ""
	hasFallback := len(args) > 1
	if hasFallback {
		fallback = strings.Join(args[1:], ", ")
	}
	if !strings.HasPrefix(key, "--") {
		return "", Errorf("The --theme(…) function can only be used with CSS variables from your theme, but received `%s`.", key)
	}
	resolved, ok := ds.Theme.ResolveThemeValue(key, inline)
	if !ok {
		if hasFallback {
			return fallback, nil
		}
		return "", Errorf("Could not resolve value for theme function: `--theme(%s)`.", key)
	}
	if !hasFallback || fallback == "initial" {
		return resolved, nil
	}
	if resolved == "initial" {
		return fallback, nil
	}
	if strings.HasPrefix(resolved, "var(") || strings.HasPrefix(resolved, "theme(") || strings.HasPrefix(resolved, "--theme(") {
		return InjectFallback(resolved, fallback), nil
	}
	return resolved, nil
}

func legacyThemeFunction(nodes []ValueNode, ds *DesignSystem) (string, error) {
	args := functionArgs(nodes)
	key := ""
	if len(args) > 0 {
		key = args[0]
	}
	if unquoted, ok := Unquote(key); ok {
		key = unquoted
	}
	resolved, ok := ds.Theme.ResolveThemeValue(key, true)
	if !ok {
		if len(args) > 1 {
			return strings.Join(args[1:], ", "), nil
		}
		return "", Errorf("Could not resolve value for theme function: `theme(%s)`.", key)
	}
	return resolved, nil
}

func isThemeReference(name string) bool { return name == "var" || name == "theme" || name == "--theme" }

// InjectFallback injects fallback into the innermost var(...), theme(...),
// or --theme(...) call that has no fallback or whose fallback is `initial`.
func InjectFallback(value, fallback string) string {
	ast := ParseValue(value)
	injectFallback(ast, fallback)
	return ValueToCSS(ast)
}

func injectFallback(nodes []ValueNode, fallback string) bool {
	for _, n := range nodes {
		f, ok := n.(*ValueFunction)
		if !ok {
			continue
		}
		if injectFallback(f.Nodes, fallback) {
			return true
		}
		if !isThemeReference(f.Name) {
			continue
		}
		comma := -1
		for i, child := range f.Nodes {
			if s, ok := child.(*ValueSeparator); ok && s.Value == "," {
				comma = i
				break
			}
		}
		if comma == -1 {
			f.Nodes = append(f.Nodes, Sep(","), Sep(" "), Word(fallback))
			return true
		}
		if strings.TrimSpace(ValueToCSS(f.Nodes[comma+1:])) == "initial" {
			f.Nodes = append(append([]ValueNode{}, f.Nodes[:comma+1]...), Sep(" "), Word(fallback))
			return true
		}
	}
	return false
}
